### PR TITLE
#253 - Update TestResultReader.ts to Handle jUnit Results with Empty Suites

### LINKED ISSUE #253: TestResultReader does not work with empty suites.
Hello, 
I'm trying to use the "Report Test Results" step in VSO on a NodeJS project, but I'm running into a few issues. I'm running my tests via Mocha. I've tried both the built-in `xunit` reporter and a third party [junit reporter](https://www.npmjs.com/package/mocha-junit-reporter).  The `xunit` reporter could not find the 'assemblies' node in the XML, which I'm assuming is an extension of the xunit schema you guys expect for .net projects.

The JUnit reporter creates an empty root suite, which caused an error in `TestResultReader.ts`, where a suite could not iterate its tests. I was able to work around the empty root suite, but this error also showed up when I had nested mocha structures like this: 

```js
describe("Some Component", () => {
    describe("some aspect", () => {
        it("<specific test>", () => {
        });
        ...
    });
});
```

So, in the end, if a suite doesn't contain any tests, you should just skip it.
e.g. wrap this (https://github.com/Microsoft/vso-agent/blob/master/src/agent/testresultreader.ts#L261) with `if (rootNode.testcase) {}`.  

If I get some time later this week, I'll put together a PR for this issue.

### PR DESCRIPTION
**Pull request by darthtrevino**
This gist of this fix is basically on this line: https://github.com/Microsoft/vso-agent/compare/master...darthtrevino:master#diff-c239605f4d523f5bd8f9cd70928887daR337

The code that was executed per child node was looking a little scary, cyclomatic-complexity wise, so I did a very minor function extraction.

Closes #253

### CODE REVIEW

File: src/agent/testresultreader.ts [hunk 1]
```diff
@@ -228,39 +228,19 @@ export class ResultReader {
...
         
-        //find test case nodes in JUnit result xml
-        var testResults = [];
-
-        for (var i = 0; i < rootNode.testcase.count(); i++) {
-            var testCaseNode = rootNode.testcase.at(i);
-
+        function extractTestResult(testCaseNode): testifm.TestResultCreateModel {
+            function pickFirst(type: string) {
+                if (testCaseNode[type].count() > 1) {
```
Comment (nigurr) [inline src/agent/testresultreader.ts]: It's good to handle the scenario where  testCaseNode[type] doesn't exist.

File: src/agent/testresultreader.ts [hunk 2]
```diff
@@ -346,8 +303,42 @@ export class ResultReader {
...
+        }
+
+        if (rootNode.attributes().time) {
+            totalRunDuration = parseFloat(rootNode.attributes().time); //in seconds
+        }
+        
+        //find test case nodes in JUnit result xml
+        var testResults = [];
+
+        if (rootNode.testcase) {
```
Comment (kaadhina) [inline src/agent/testresultreader.ts]: What would be the behaviour now if there are no test case nodes in the file? Would it silently pass without publishing anything? Should we instead let the user know that something went wrong?

File: src/agent/testresultreader.ts (hunk 1 above)
Comment (darthtrevino) [inline src/agent/testresultreader.ts]: I agree; I was trying to be fairly conservative and just combined what was duplicated, but I'd be glad to add another check here.

File: src/agent/testresultreader.ts (hunk 2 above)
Comment (darthtrevino) [inline src/agent/testresultreader.ts]: This is a better edge case than the test I added, so I'll update the new test to reflect this situation. The right thing to do here would be to let the task pass and emit a warning message in the logs about there being no tests. I'm not sure what API emits log messages, but I'll dig into that a little bit.

Comment (darthtrevino) [inline src/agent/testresultreader.ts]: I've added some unit tests for these cases, but I'll have to run an integration test later to see if it's working out.

**Comment by nigurr** [comment]
@darthtrevino Can you please resolve the conflicts? Rest of the code is pretty fine now. 
Thanks for adding tests in default process.

**Comment by darthtrevino** [comment]
Aye aye. Upstream is merged.
